docs-client/src/components/VariableList/index.tsx

```diff
@@ -56,39 +65,184 @@ export default function({
             <TableCell>Required</TableCell>
             <TableCell>Type</TableCell>
             <TableCell>Description</TableCell>
+            {hasBean && <TableCell />}
           </TableRow>
         </TableHead>
         <TableBody>
-          {variables.length > 0 ? (
-            variables.map((variable) => (
-              <TableRow key={variable.name}>
+          <FieldInfos
+            indent={0}
+            title={title}
+            hasLocation={hasLocation}
+            variables={variables}
+            specification={specification}
+            hasBean={hasBean}
+            childIndex={0}
+          />
+        </TableBody>
+      </Table>
+      <Typography variant="body1" paragraph />
+    </>
+  );
+}
+
+interface State {
+  isEmpty: boolean;
+  isBeans: boolean[];
+  openBeans: boolean[];
+  colSpanLength: number;
+  childStates: { [key: number]: State };
+}
+
+interface OwnProps {
+  indent: number;
+  hasBean: boolean;
+  state?: State;
+  setChildStateInParent?: (index: number, childState: State) => void;
+  childIndex: number;
+}
+
+type FieldInfosProps = OwnProps & Props;
+
+class FieldInfos extends React.Component<FieldInfosProps, State> {
+  private static generateKey(pre: string): string {
+    return `${pre}_${new Date().getTime()}`;
+  }
+
+  private static indentString(indent: number, s: string): string {
+    return `${'\xa0'.repeat(indent)}${s}`;
+  }
+
+  constructor(props: FieldInfosProps) {
+    super(props);
+
+    this.setChildStateInParent = this.setChildStateInParent.bind(this);
+
+    if (props.state) {
+      this.state = props.state;
+      return;
+    }
+
+    const fieldInfos = props.variables;
+
+    let isEmpty;
+    let isBeans: boolean[] = [];
+    let openBeans: boolean[] = [];
+    if (fieldInfos.length > 0) {
```

Comment:
How about setting`isEmpty` first and using it with conditions, e.g.
( I think it would be more clear to read. )

```typescript
let isEmpty = fieldInfos.length == 0;
...
if (!isEmpty) {
...
} else {
...
}
```